### PR TITLE
Add exhaustive tuner

### PR DESCRIPTION
**Pull request by edumucelli**
Hi,

This PR adds an exhaustive tuner, i.e., a Tuner that explores all the trial space. I needed it and I saw that others too (https://github.com/keras-team/keras-tuner/issues/192 and https://github.com/keras-team/keras-tuner/issues/408), but unfortunately there was not a way of doing that given `RandomSearch` has an internal `max_collisions` that will stop the search if too many of the same parameter combination has already been explored.

The change is self contained, everything goes into a single file. I am using it to train a 600+ combination space. The main limitation is that it only allows `hp.Choice`, but I am open to improve it. Adding `hp.Boolean`, `hp.Fixed`, or even using the step on `hp.Int` to generate all the combinations before-hand for all the types.

I will surely add some tests if you guys are positive about this change.

### CODE REVIEW

**Review by haifeng-jin — COMMENTED** [review]
Thank you so much for the PR! This is really an important feature that we haven't got a chance to implement.
However, as this is a big feature. We need to go through some approval process.
Please be patient with us. : )

**Comment by edumucelli** [comment]
> Thank you so much for the PR! This is really an important feature that we haven't got a chance to implement. However, as this is a big feature. We need to go through some approval process. Please be patient with us. : )
> 
> I will likely to pull request to your branch for some changes.

Hi @haifeng-jin, thanks from the prompt reply and great you are positive about it. I completely understand the timing needed to release such feature.

I can take it to completion, even if requires improvements as I imagine, just guide me through what is necessary and I can make it happen. The improvements I imagine are mandatory: 1) tests, I have seen how they are done for other tunes 2) adding support to other hyperparameters types.

Let me know.

**Comment by edumucelli** [comment]
I have added a first test to cover the basics of it.

**Comment by haifeng-jin** [comment]
Hi @edumucelli ,

Here is what will happen next.

I will initiate a doc for internal approval, which contains the motivation for exhaustive tuner, user code examples of using exhaustive tuner, the description of the implementation and arugments and so on.

I believe, we already have most of them in this PR.

As I am on parental leave, my working time is limited.
Expect me to come back with the needed changes to this PR in 3 weeks.

Thanks.

**Comment by edumucelli** [comment]
Hi @haifeng-jin, thanks for the information. Very clear.
No problems, enjoy your time off.
We keep it going in 3 weeks.

**Comment by edumucelli** [comment]
One more thing: I will allow it to take into account `hp.Fixed` and `hp.Bool`. As far as I can see those could be supported with no extra code compared to the existing PR and add widens the scope of the tuner.

**Comment by haifeng-jin** [comment]
@edumucelli
Internal review finished.
We will merge your pull request after making some changes. Thank you!

The easiest way is for me to make the changes directly.
I can add a few commits and you can take a look afterwards.
Would you please follow [these instructions](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to allow me to commit to your branch?

The changes may include:
* Renaming the tuner to `GridSearch`.
* Add the `max_trial` argument, as people may need to use it to limit the number of trials.
* Refactor the hyperparameters classes a little bit to support all types of hyperparameters in this tuner.
* Change the implementation of the oracle from "keep random sample until find all" to an iterator style method to go through all possible combinations in order.

**Comment by edumucelli** [comment]
> @edumucelli Internal review finished. We will merge your pull request after making some changes. Thank you!
> 

Great!

> The easiest way is for me to make the changes directly. I can add a few commits and you can take a look afterwards. Would you please follow [these instructions](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to allow me to commit to your branch?
I will take a look.
> 
> The changes may include:
> 
>     * Renaming the tuner to `GridSearch`.
Just did it
>     * Add the `max_trial` argument, as people may need to use it to limit the number of trials.
I can do that one as well.
>     * Refactor the hyperparameters classes a little bit to support all types of hyperparameters in this tuner. This needs to be done after [Improve `Int` and `Float` functionality, docstirngs and code structure #725](https://github.com/keras-team/keras-tuner/pull/725) .
>     * Change the implementation of the oracle from "keep random sample until find all" to an iterator style method to go through all possible combinations in order.
I can probably do that one.
>     * May also support `conditional_scope` if find it not hard to implement. If too hard to implement, it will be in a follow-up PR.
With enough guidance I can do that, but AFAIK conditional scope is for hyperparameters, how would that be related to GridSearch?
>     * Improve test coverage for the added code to be 100% covered.
I can do that too.

**Comment by haifeng-jin** [comment]
@edumucelli 

Here are the docstrings we revised.
You may start from here to update the function definitions and code accordingly.

```py
class GridSearchOracle(...):
    """Grid search oracle.

    Args:
        objective: A string, `keras_tuner.Objective` instance, or a list of
            `keras_tuner.Objective`s and strings. If a string, the direction of
            the optimization (min or max) will be inferred. If a list of
            `keras_tuner.Objective`, we will minimize the sum of all the
            objectives to minimize subtracting the sum of all the objectives to
            maximize. The `objective` argument is optional when
            `Tuner.run_trial()` or `HyperModel.fit()` returns a single float as
            the objective to minimize.
        max_trials: Optional integer, the total number of trials (model
            configurations) to test at most. Note that the oracle may interrupt
            the search before `max_trial` models have been tested if the search
            space has been exhausted. If left unspecified, it runs till the
            search space is exhausted.
        seed: Optional integer, the random seed.
        hyperparameters: Optional `HyperParameters` instance. Can be used to
            override (or register in advance) hyperparameters in the search
            space.
        tune_new_entries: Boolean, whether hyperparameter entries that are
            requested by the hypermodel but that were not specified in
            `hyperparameters` should be added to the search space, or not. If
            not, then the default value for these parameters will be used.
            Defaults to True.
        allow_new_entries: Boolean, whether the hypermodel is allowed to
            request hyperparameter entries not listed in `hyperparameters`.
            Defaults to True.
    """
```

```py
class GridSearch(...):
    """The Grid search tuner.
    
    It will try all the possible hyperparameter
    combinations up to exhaustion.

    For example:

    ```py
    optimizer = hp.Choice("model_name", values=["sgd", "adam"])
    lr = hp.Choice("lr", values=[0.01, 0.1])
    ```

    This tuner will fit models for: ["sgd", 0.01], ["sgd", 0.1], ["adam", 0.01]
    ["adam", 0.1].

    Args:
        hypermodel: Instance of `HyperModel` class (or callable that takes
            hyperparameters and returns a Model instance). It is optional when
            `Tuner.run_trial()` is overridden and does not use
            `self.hypermodel`.
        objective: A string, `keras_tuner.Objective` instance, or a list of
            `keras_tuner.Objective`s and strings. If a string, the direction of
            the optimization (min or max) will be inferred. If a list of
            `keras_tuner.Objective`, we will minimize the sum of all the
            objectives to minimize subtracting the sum of all the objectives to
            maximize. The `objective` argument is optional when
            `Tuner.run_trial()` or `HyperModel.fit()` returns a single float as
            the objective to minimize.
        max_trials: Optional integer, the total number of trials (model
            configurations) to test at most. Note that the oracle may interrupt
            the search before `max_trial` models have been tested if the search
            space has been exhausted. If left unspecified, it runs till the
            search space is exhausted.
        seed: Optional integer, the random seed.
        hyperparameters: Optional `HyperParameters` instance. Can be used to
            override (or register in advance) hyperparameters in the search
            space.
        tune_new_entries: Boolean, whether hyperparameter entries that are
            requested by the hypermodel but that were not specified in
            `hyperparameters` should be added to the search space, or not. If
            not, then the default value for these parameters will be used.
            Defaults to True.
        allow_new_entries: Boolean, whether the hypermodel is allowed to
            request hyperparameter entries not listed in `hyperparameters`.
            Defaults to True.
        **kwargs: Keyword arguments relevant to all `Tuner` subclasses.
            Please see the docstring for `Tuner`.
    """
```

**Comment by edumucelli** [comment]
@haifeng-jin added:
* Revised docstring
* `max_trials`, I will add a specific test for it covering a scenario in which although the hps allow more trials, the max_trials limit.

**Comment by haifeng-jin** [comment]
These 3 items should be done together.
```
* Refactor the hyperparameters classes a little bit to support all types of hyperparameters in this tuner. This needs to be done after [Improve `Int` and `Float` functionality, docstirngs and code structure #725](https://github.com/keras-team/keras-tuner/pull/725) .
* Change the implementation of the oracle from "keep random sample until find all" to an iterator style method to go through all possible combinations in order.
* May also support `conditional_scope` if find it not hard to implement. If too hard to implement, it will be in a follow-up PR.
```

The `conditional_scope` can decide which hps are active.
The tuner need to be aware of this. Otherwise, it would keep iterating the inactive ones, which has no influence on the model building.

```py
# First add a method to all `HyperParameter` subclasses
# to get all possible values.
class Int(...):
    @property
    def all_values(self):
        ...
        return all_possible_values

class GridSearchOracle(...):
    def __init__(self, ...):
        # key is hp name, value is the last hp value
        self.last_values = dict()
        # key is hp name, value is boolean
        # marking it was active or not last trial.
        self.last_active = dict()

    def populate_space(self, ...):
        values = self._next_combination()
        self.last_values = values
        # Also update the self.last_active
        ...
    
    def _next_combination(self):
        # Find the next combination based on the last values 
        # and last active state and the `all_values` method.
        
        # It is like reimplementing itertools.product, but 
        # take whether is active into account.
```

I will take over to implement this after you finished your part.

Thanks!

File: keras_tuner/tuners/gridsearch.py
Comment (haifeng-jin) [inline keras_tuner/tuners/gridsearch.py:91]: We should move the logics you added in `create_trial` into `populate_space` to do them on the fly.
Prefer to keep `create_trial` untouched unless we really need to.

**Review by haifeng-jin — CHANGES_REQUESTED** [review]

**Comment by haifeng-jin** [comment]
@edumucelli I am taking over from now and you may take a look later. Thank you!

**Pull request merged**
